File: rust/src/equipment/latex_table.rs
use crate::core_mechanics::abilities::replace_attack_terms;
use crate::equipment::{item_creature, rank_and_price_text, ItemRarity, StandardItem};
use crate::latex_formatting::latexify;

pub trait ToTableRows {
    fn to_table_rows(&self) -> Vec<TableRow>;
}

#[derive(Clone)]
pub struct TableRow {
    category: Option<String>,
    consumable: bool,
    name: String,
    magical: bool,
    rank: i32,
    rarity: ItemRarity,
    short_description: String,
}

impl TableRow {
    pub fn from_item(item: &StandardItem, consumable: bool, category: Option<String>) -> Vec<Self> {
        let mut rows = vec![];
        rows.push(Self {
            category: category.clone(),
            consumable,
            magical: item.magical,
            name: item.name.clone(),
            rank: item.rank,
            rarity: item.rarity.clone(),
            short_description: item.short_description.clone(),
        });

        for upgraded_item in item.upgrade_items() {
            rows.push(Self {
                category: category.clone(),
                consumable,
                magical: item.magical,
                name: upgraded_item.name,
                rank: upgraded_item.rank,
                rarity: upgraded_item.rarity.clone(),
                short_description: upgraded_item.short_description,
            });
        }

        rows
    }

    fn to_latex(&self, percentile: Option<&str>) -> String {
        let latex = format!(
            "
                \\itemref<{name}>{sparkle} {category_separator} {category}
                & {short_description}
                & {rank_and_price}
                & {page_or_percentile}
                \\\\
            ",
            name = self.name,
            sparkle = if self.magical { r"\sparkle" } else { "" },
            category_separator = if self.category.is_some() { "&" } else { "" },
            category = self.category.clone().unwrap_or(String::from("")),
            rank_and_price = rank_and_price_text(self.rank, &self.rarity),
            short_description = self.short_description.trim(),
            page_or_percentile = percentile.unwrap_or(&format!("\\itempref<{}>", self.name)),
        );
        replace_attack_terms(latex.trim(), &item_creature(self.rank), false, None)
    }
}

fn table_caption(caption: &str) -> String {
    format!("\\lcaption<{caption}> \\\\")
}

pub fn table_header(name_text: &str, with_category: bool, with_percentile: bool) -> String {
    format!(
        "
            \\tb<{name}>{category_separator} {category_column_name} & \\tb<Description> & \\tb<Rank (Cost)> & {page_or_percentile} \\tableheaderrule
        ",
        name = name_text,
        category_separator = if with_category { "&" } else { "" },
        // TODO: are there other reasonable category column names?
        category_column_name = if with_category { r"\tb{Type}" } else {""},
        page_or_percentile = if with_percentile { r"\tb{d100}" } else { r"\tb{Page}" },
    )
}

pub fn standard_sort(rows: &mut Vec<TableRow>) {
    // Primary sort is by rank, secondary sort is by consumability, tertiary sort is by category,
    // final sort is by name.
    rows.sort_by(|a, b| a.name.cmp(&b.name));
    rows.sort_by(|a, b| a.category.cmp(&b.category));
    // Flip the boolean so consumables come before permanent items
    rows.sort_by(|a, b| (!a.consumable).cmp(&!b.consumable));
    rows.sort_by(|a, b| a.rank.cmp(&b.rank));
}

pub fn longtable(caption: &str, rows: Vec<TableRow>, with_category: bool) -> String {
    latexify(format!(
        "
            \\begin<longtablewrapper>
            \\begin<longtable><p<17em> {category_and_effects} p<6em> p<3em>>
                {caption}
                {header}
                {rows}
            \\end<longtable>
            \\end<longtablewrapper>
        ",
        category_and_effects = if with_category {
            "p{5em} p{20em}"
        } else {
            "p{26em}"
        },
        caption = table_caption(caption),
        header = table_header("Name", with_category, false),
        rows = rows
            .iter()
            .map(|r| r.to_latex(None))
            .collect::<Vec<String>>()
            .join("\n"),
    ))
}

// Like a TableRow, but has a Percentile column. That indicates the roll on a d100 required to
// get that item from a random loot drop.
struct PercentileTableRow {
    percentile_range: String,
    table_row: TableRow,
}

impl PercentileTableRow {
    fn to_latex(&self) -> String {
        self.table_row.to_latex(Some(&self.percentile_range))
    }
}

// Like longtable, but instead of having a "pages" column, it has a "d100" column.
// Useful for random loot drops.
pub fn longtable_percentile(caption: &str, rows: Vec<TableRow>, with_category: bool) -> String {
    // Convert the TableRows to PercentileTableRows and assign each row a percentile_range based on
    // its position within its rank.
    let mut percentile_rows: Vec<PercentileTableRow> = vec![];
    for rank in -1..9 {
        let rows_at_rank: Vec<&TableRow> = rows.iter().filter(|r| r.rank == rank).collect();
        let distinct_item_count = rows_at_rank.len();
        if distinct_item_count == 0 {
            continue;
        }
        // Rolling a 100 upgrades to the next rank, so we don't include it in the percentile
        // listing.
        let step_size = 99.0 / distinct_item_count as f64;
        let mut min_roll = 0;
        // We keep this as an f64 so it can have smooth steps.
        let mut true_max_roll = step_size;
        for row in rows_at_rank {
            let max_roll_int = true_max_roll.round() as i32;
            percentile_rows.push(PercentileTableRow {
                percentile_range: if max_roll_int == min_roll {
                    min_roll.to_string()
                } else {
                    format!("{}--{}", min_roll, max_roll_int)
                },
                table_row: row.clone(),
            });
            min_roll = 1 + max_roll_int;
            true_max_roll += step_size;
        }
    }

    latexify(format!(
        "
            \\begin<longtablewrapper>
            \\begin<longtable><p<17em> {category_and_effects} p<6em> p<3em>>
                {caption}
                {header}
                {rows}
            \\end<longtable>
            \\end<longtablewrapper>
        ",
        category_and_effects = if with_category {
            "p{5em} p{20em}"
        } else {
            "p{26em}"
        },
        caption = table_caption(caption),
        header = table_header("Name", with_category, true),
        rows = percentile_rows
            .iter()
            .map(|r| r.to_latex())
            .collect::<Vec<String>>()
            .join("\n")
    ))
}

File: typescript/src/equipment/latex/latex_table.ts
import { ItemRarity, StandardItem, getUpgradeItems } from '../types';
import { getRankAndPriceText } from '../item_price';
import { getItemCreature } from '../item_creature';
import { replacePlaceholders } from '../../latex/monsters/replace_placeholders';
import { latexify } from '../../latex/format';

export interface TableRow {
  category?: string;
  consumable: boolean;
  name: string;
  magical: boolean;
  rank: number;
  rarity: ItemRarity;
  short_description: string;
  tags?: string[];
}

export function fromItem(item: StandardItem, consumable: boolean, category?: string): TableRow[] {
  const rows: TableRow[] = [];
  rows.push({
    category,
    consumable,
    magical: item.magical,
    name: item.name,
    rank: item.rank,
    rarity: item.rarity,
    short_description: item.short_description,
    tags: item.tags,
  });

  for (const upgradedItem of getUpgradeItems(item)) {
    rows.push({
      category,
      consumable,
      magical: item.magical,
      name: upgradedItem.name,
      rank: upgradedItem.rank,
      rarity: upgradedItem.rarity,
      short_description: upgradedItem.short_description,
      tags: upgradedItem.tags,
    });
  }

  return rows;
}

export function rowToLatex(
  row: TableRow,
  percentile?: string,
  withAttunement: boolean = false,
): string {
  const sparkle = row.magical ? '\\sparkle' : '';
  const categorySeparator = row.category ? '&' : '';
  const category = row.category || '';

  const attunementSeparator = withAttunement ? '&' : '';
  let attunementText = '';
  if (withAttunement) {
    const tags = row.tags || [];
    if (tags.some((t) => t.toLowerCase() === 'attune (deep)')) {
      attunementText = 'Deep';
    } else if (tags.some((t) => t.toLowerCase() === 'attune')) {
      attunementText = 'Yes';
    } else {
      attunementText = 'No';
    }
  }

  const rankAndPrice = getRankAndPriceText(row.rank, row.rarity);
  const pageOrPercentile = percentile || `\\itempref{${row.name}}`;

  const latex = `
    \\itemref{${row.name}}${sparkle} ${categorySeparator} ${category} ${attunementSeparator} ${attunementText}
    & ${row.short_description.trim()}
    & ${rankAndPrice}
    & ${pageOrPercentile}
    \\\\
  `.trim();

  const creature = getItemCreature(row.rank);
  return replacePlaceholders(creature, latex, { isMagical: row.magical });
}

function tableCaption(caption: string): string {
  return `\\lcaption{${caption}} \\\\`;
}

export function tableHeader(
  nameText: string,
  withCategory: boolean,
  withPercentile: boolean,
  withAttunement: boolean = false,
): string {
  const categorySeparator = withCategory ? '&' : '';
  const categoryColumnName = withCategory ? ' \\tb{Type}' : '';
  const attunementSeparator = withAttunement ? '&' : '';
  const attunementColumnName = withAttunement ? ' \\tb{Attunement}' : '';
  const pageOrPercentile = withPercentile ? '\\tb{d100}' : '\\tb{Page}';

  return `
    \\tb{${nameText}}${categorySeparator}${categoryColumnName}${attunementSeparator}${attunementColumnName} & \\tb{Description} & \\tb{Rank (Cost)} & ${pageOrPercentile} \\tableheaderrule
  `.trim();
}

export function standardSort(rows: TableRow[]): void {
  rows.sort((a, b) => {
    // Primary: Rank
    if (a.rank !== b.rank) return a.rank - b.rank;
    // Secondary: Consumability (consumables first)
    if (a.consumable !== b.consumable) return a.consumable ? -1 : 1;
    // Tertiary: Category
    if (a.category !== b.category) {
      const catA = a.category || '';
      const catB = b.category || '';
      return catA < catB ? -1 : 1;
    }
    // Final: Name
    return a.name < b.name ? -1 : a.name > b.name ? 1 : 0;
  });
}

export interface LongtableOptions {
  caption: string;
  rows: TableRow[];
  withCategory: boolean;
  withAttunement?: boolean;
}

export function longtable(options: LongtableOptions): string {
  const { caption, rows, withCategory, withAttunement = false } = options;
  let colSpec = '';
  if (withCategory && withAttunement) {
    colSpec = 'p{17em} p{5em} p{5em} p{15em} p{6em} p{3em}';
  } else if (withCategory) {
    colSpec = 'p{17em} p{5em} p{20em} p{6em} p{3em}';
  } else if (withAttunement) {
    colSpec = 'p{17em} p{5em} p{21em} p{6em} p{3em}';
  } else {
    colSpec = 'p{17em} p{26em} p{6em} p{3em}';
  }

  const content = `
\\begin{longtablewrapper}
\\begin{longtable}{${colSpec}}
    ${tableCaption(caption)}
    ${tableHeader('Name', withCategory, false, withAttunement)}
    ${rows.map((r) => rowToLatex(r, undefined, withAttunement)).join('\n')}
\\end{longtable}
\\end{longtablewrapper}
  `.trim();

  return latexify(content);
}

export function longtablePercentile({
  caption,
  rows,
  withCategory,
  withAttunement = false,
}: LongtableOptions): string {
  const percentileRows: { range: string; row: TableRow }[] = [];

  for (let rank = -1; rank <= 8; rank++) {
    const rowsAtRank = rows.filter((r) => r.rank === rank);
    const distinctItemCount = rowsAtRank.length;
    if (distinctItemCount === 0) continue;

    const stepSize = 99.0 / distinctItemCount;
    let minRoll = 0;
    let trueMaxRoll = stepSize;

    for (const row of rowsAtRank) {
      const maxRollInt = Math.round(trueMaxRoll);
      percentileRows.push({
        range: maxRollInt === minRoll ? `${minRoll}` : `${minRoll}--${maxRollInt}`,
        row: row,
      });
      minRoll = 1 + maxRollInt;
      trueMaxRoll += stepSize;
    }
  }

  let colSpec = '';
  if (withCategory && withAttunement) {
    colSpec = 'p{17em} p{5em} p{5em} p{15em} p{6em} p{3em}';
  } else if (withCategory) {
    colSpec = 'p{17em} p{5em} p{20em} p{6em} p{3em}';
  } else if (withAttunement) {
    colSpec = 'p{17em} p{5em} p{21em} p{6em} p{3em}';
  } else {
    colSpec = 'p{17em} p{26em} p{6em} p{3em}';
  }

  const content = `
\\begin{longtablewrapper}
\\begin{longtable}{${colSpec}}
    ${tableCaption(caption)}
    ${tableHeader('Name', withCategory, true, withAttunement)}
    ${percentileRows.map((pr) => rowToLatex(pr.row, pr.range, withAttunement)).join('\n')}
\\end{longtable}
\\end{longtablewrapper}
  `.trim();

  return latexify(content);
}
